bindings/feishu/webhook/webhook.go

```diff
@@ -0,0 +1,224 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation and Dapr Contributors.
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+
+package webhook
+
+import (
+	"bytes"
+	"context"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/dapr/components-contrib/bindings"
+	"github.com/dapr/kit/logger"
+)
+
+const (
+	webhookContentType       = "application/json"
+	defaultHTTPClientTimeout = time.Second * 30
+)
+
+type FeishuWebhook struct {
+	logger     logger.Logger
+	settings   Settings
+	httpClient *http.Client
+}
+
+type webhookResult struct {
+	Code int    `json:"code"`
+	Msg  string `json:"msg"`
+}
+
+type outgoingWebhook struct {
+	handler func(*bindings.ReadResponse) ([]byte, error)
+}
+
+var webhooks = struct { // nolint: gochecknoglobals
+	sync.RWMutex
+	m map[string]*outgoingWebhook
+}{m: make(map[string]*outgoingWebhook)}
+
+func NewFeishuWebhook(l logger.Logger) *FeishuWebhook {
+	// See guidance on proper HTTP client settings here:
+	// https://medium.com/@nate510/don-t-use-go-s-default-http-client-4804cb19f779
+	dialer := &net.Dialer{ //nolint:exhaustivestruct
+		Timeout: 5 * time.Second,
+	}
+	var netTransport = &http.Transport{ //nolint:exhaustivestruct
+		DialContext:         dialer.DialContext,
+		TLSHandshakeTimeout: 5 * time.Second,
+	}
+	httpClient := &http.Client{ //nolint:exhaustivestruct
+		Timeout:   defaultHTTPClientTimeout,
+		Transport: netTransport,
+	}
+
+	return &FeishuWebhook{ //nolint:exhaustivestruct
+		logger:     l,
+		httpClient: httpClient,
+	}
+}
+
+// Init performs metadata parsing
+func (t *FeishuWebhook) Init(metadata bindings.Metadata) error {
+	var err error
+	if err = t.settings.Decode(metadata.Properties); err != nil {
+		return fmt.Errorf("feishu configuration error: %w", err)
+	}
+	if err = t.settings.Validate(); err != nil {
+		return fmt.Errorf("feishu configuration error: %w", err)
+	}
+
+	return nil
+}
+
+// Read triggers the outgoing webhook, not yet production ready
+func (t *FeishuWebhook) Read(handler func(*bindings.ReadResponse) ([]byte, error)) error {
+	t.logger.Debugf("Feishu webhook: start read input binding")
+
+	webhooks.Lock()
+	defer webhooks.Unlock()
+	_, loaded := webhooks.m[t.settings.ID]
+	if loaded {
+		return fmt.Errorf("feishu webhook error: duplicate id %s", t.settings.ID)
+	}
+	webhooks.m[t.settings.ID] = &outgoingWebhook{handler: handler}
```

Comment:
It looks like `Read` (Input Binding method) must be called for a webhook to be created so that `Invoke` with `GetOperation` (Output Binding method) can be called to retrieve a message . I don't think this is how this is intended to work. Is there a reason why you set it up this way? Maybe we can recommend a different approach that fits into the model.

/cc: @artursouza @mukundansundar Thoughts?

Reply:
Any reason why this cannot go to Init()?

Reply:
Looks like this was borrowed from the DingTalk component and the same questions apply to that component.
https://github.com/dapr/components-contrib/blob/master/bindings/alicloud/dingtalk/webhook/webhook.go#L88-L101

Reply:
@pkedy @artursouza yes, it copied from DingTalk component, I will update this later

Reply:
@ipfans can you please update?

Reply:
@yaron2 @pkedy Sorry, I had almost forgot it.Here is the determination if there are duplicate Webhook ID, is there any alternative solution to achieve a similar goal?